src/ai/backend/common/api_handlers.py

```diff
@@ -0,0 +1,348 @@
+
+
+class BodyParam(Generic[T]):
+    _model: Type[T]
+    _parsed: Optional[T]
+
+    def __init__(self, model: Type[T]) -> None:
+        self._model = model
+        self._parsed: Optional[T] = None
+
+    @property
+    def parsed(self) -> T:
+        if not self._parsed:
+            raise ParameterNotParsedError(
+                f"Parameter of type {self._model.__name__} has not been parsed yet"
+            )
+        return self._parsed
+
+    def from_body(self, json_body: str) -> Self:
+        self._parsed = self._model.model_validate(json_body)
+        return self
+
+
+class QueryParam(Generic[T]):
+    _model: Type[T]
+    _parsed: Optional[T]
+
+    def __init__(self, model: Type[T]) -> None:
+        self._model = model
+        self._parsed: Optional[T] = None
+
+    @property
+    def parsed(self) -> T:
+        if not self._parsed:
+            raise ParameterNotParsedError(
+                f"Parameter of type {self._model.__name__} has not been parsed yet"
+            )
+        return self._parsed
+
+    def from_query(self, query: MultiMapping[str]) -> Self:
+        self._parsed = self._model.model_validate(query)
+        return self
+
+
+class HeaderParam(Generic[T]):
+    _model: Type[T]
+    _parsed: Optional[T]
+
+    def __init__(self, model: Type[T]) -> None:
+        self._model = model
+        self._parsed: Optional[T] = None
+
+    @property
+    def parsed(self) -> T:
+        if not self._parsed:
+            raise ParameterNotParsedError(
+                f"Parameter of type {self._model.__name__} has not been parsed yet"
+            )
+        return self._parsed
+
+    def from_header(self, headers: CIMultiDictProxy[str]) -> Self:
+        self._parsed = self._model.model_validate(headers)
+        return self
+
+
+class PathParam(Generic[T]):
+    _model: Type[T]
+    _parsed: Optional[T]
+
+    def __init__(self, model: Type[T]) -> None:
+        self._model = model
+        self._parsed: Optional[T] = None
+
+    @property
+    def parsed(self) -> T:
+        if not self._parsed:
+            raise ParameterNotParsedError(
+                f"Parameter of type {self._model.__name__} has not been parsed yet"
+            )
+        return self._parsed
+
+    def from_path(self, match_info: UrlMappingMatchInfo) -> Self:
+        self._parsed = self._model.model_validate(match_info)
+        return self
+
+
+class MiddlewareParam(ABC, BaseModel):
+    @classmethod
+    @abstractmethod
+    def from_request(cls, request: web.Request) -> Self:
+        pass
+
+
+class BaseResponseModel(BaseModel):
+    pass
+
+
+JSONDict: TypeAlias = dict[str, Any]
+
+
+@dataclass
+class APIResponse:
+    _status_code: int
+    _data: Optional[BaseResponseModel]
+
+    @classmethod
+    def build(cls, status_code: int, response_model: BaseResponseModel) -> Self:
+        return cls(_status_code=status_code, _data=response_model)
+
+    @classmethod
+    def no_content(cls, status_code: int) -> Self:
```

Comment:
The name `no_content` is confusing. I'd suggest `empty`.
Refs
- https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/ResponseEntity.html#noContent()
- https://learn.microsoft.com/ko-kr/dotnet/api/microsoft.aspnetcore.http.httpresults.nocontent?view=aspnetcore-9.0

Reply:
Alternatively, you could allow `None` for `data` in `build()` and remove `no_content()`.